src/eids.js

```diff
@@ -195,3 +196,61 @@ export class EidRewriter {
     return this.ensureUnique(`${eId}_${count}`, false);
   }
 }
+
+export class WorkComponentRewriter {
+  constructor () {
+    this.counters = {};
+    this.skipElements = ['meta', 'preface', 'preamble', 'body', 'mainBody'];
+  }
+
+  /** Rewrites the FRBR URI work components for all attachment nodes in the tree.
+   */
+  rewriteAllAttachmentWorkComponents (element, prefix = '') {
+    this.counters = {};
+    this.rewriteAttachmentWorkComponent(element, prefix);
+  }
+
+  rewriteAttachmentWorkComponent (element, prefix = '') {
+    let tag = element.tagName;
+    if (this.skipElements.includes(tag)) {
+      return;
+    }
+
+    if (tag === 'attachment') {
+      let newName = this.getComponentName(element, prefix),
+          workComponents = element.getElementsByTagName('FRBRthis');
```

Comment:
if we could use xpath here instead we could grab just the first level. as is, we'd be rewriting the work components for all nested attachments as well, and then changing them (back) when we get down to that level, which is a bit silly.

Reply:
In practice I don't think bluebell supports nested attachments, so this is probably academic and unlikely, so I think this is fine and you don't need to optimise it further.

But you can use xpath if you want.

Reply:
It definitely does support them 😁 — but I looked into xpath and it just doesn't seem worth the effort for a pretty rare case that'll be immediately fixed anyway

Reply:
Lol, how good is bluebell?! 💪 

I agree.